src/clegoues/genprog4java/Search/Search.java

```diff
@@ -263,41 +263,51 @@ public void mutate(Representation<G> variant) throws GiveUpException {
 
 		ArrayList<Location> faultyAtoms = localization.getFaultLocalization();
 		
-		ArrayList<Location> proMutList = new ArrayList<Location>();
 		boolean foundMutationThatCanApplyToAtom = false;
 		boolean alreadySetAllStmtsToFixLoc = false;
 		while(!foundMutationThatCanApplyToAtom){
+			ArrayList<Location> proMutList = new ArrayList<Location>();
 			//promut default is 1 // promut stands for proportional mutation rate, which controls the probability that a genome is mutated in the mutation step in terms of the number of genes within it should be modified.
 			for (int i = 0; i < Search.promut; i++) {
 				//chooses a random location
-				Pair<?, Double> wa = null;
-				boolean alreadyOnList = false;
+				Location wa = null;
+				boolean foundUsableLoc = false;
+
 				//If it already picked all the fix atoms from current FixLocalization, then start picking from the ones that remain
 				if(proMutList.size()>=faultyAtoms.size()){ 
 					localization.setAllPossibleStmtsToFixLocalization();				
 					alreadySetAllStmtsToFixLoc = true;
+					faultyAtoms = localization.getFaultLocalization();
```

Comment:
Not sure if this was the right thing to do here. I'm not entirely sure how `localization.setAllPossibleStmtsToFixLocation` actually affects the fault localization; looking at the code, it seems that method affects a field called `this.fixLocations` or something, but the way we used to query for locations, `localization.getRandomLocation`, uses `this.faultLocalization`